README.md
# MetadataLogger

Logging with metadata.

## Installation

The package can be installed by adding `metadata_logger` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [{:metadata_logger, "~> 0.2"}]
end
```

## Configuration

```elixir
config :logger, :console,
  level: :debug,
  format: {MetadataLogger, :format},
  colors: [enabled: false],
  metadata: :all,
  truncate: :infinity,
  utc_log: true
```

Full documentation can be found at [https://hexdocs.pm/metadata_logger](https://hexdocs.pm/metadata_logger).


CHANGELOG.md
# Changelog

## [Unreleased]

## 0.3.2 - 2022-06-04

### Fixed

- Add a newline when producing a formatting error message (#19 by @jbruggem)

## 0.3.1 - 2020-09-09

### Fixed

- Ignore `error_logger` metadata.

## 0.3.0 - 2020-07-17

### Fixed

- Handle Elixir logger metadata: `domain`, `callers`, `ancestors`

### Changed

- Produce JSON when formating failed

## 0.2.1 - 2020-01-28

### Fixed

- Handle erlang logger metadata: `gid`, `mfa`, `report_cb`

## 0.2.0 - 2019-11-14

### Added

- `MetadataLogger.log_to_map/4`

### Changed

- Renamed from `metadata_logger_json_formatter` to `metadata_logger`.
- `MetadataLogger.format/4` now returns IO data.

## 0.1.0 - 2018-08-08

[Unreleased]: https://github.com/elixir-metadata-logger/metadata_logger/compare/v0.3.1...HEAD


test/test_helper.exs
ExUnit.start()
Logger.App.start()


test/metadata_logger_test.exs
defmodule MetadataLoggerTest do
  use ExUnit.Case
  doctest MetadataLogger

  @ts_tuple {{2019, 11, 22}, {12, 23, 45, 678}}
  @ts_iso8601 "2019-11-22T12:23:45.000678"

  test "uses to_string for struct message" do
    expected = %{
      "module" => "Elixir.MetadataLogger",
      "pid" => "#PID<" <> _ = inspect(self()),
      "metadata" => %{"foo" => "bar", "list" => [1, 2, 3]},
      "timestamp" => @ts_iso8601,
      "level" => "info",
      "message" => "https://elixir-lang.org"
    }

    got =
      parse_formatted(:info, URI.parse("https://elixir-lang.org"), @ts_tuple,
        module: MetadataLogger,
        function: "hello/1",
        file: "/my/file.ex",
        line: 11,
        pid: self(),
        foo: :bar,
        list: [1, 2, 3]
      )

    assert expected == got
  end

  test "cannot handle a map with non-encodable" do
    got =
      parse_formatted(:info, %{uri: URI.parse("https://elixir-lang.org")}, @ts_tuple,
        module: MetadataLogger,
        function: "hello/1",
        file: "/my/file.ex",
        line: 11,
        pid: self(),
        foo: :bar,
        list: [1, 2, 3]
      )
      |> Map.fetch!("message")

    assert String.starts_with?(got, "could not format json message")
  end

  test "moves known metadata into top level" do
    expected = %{
      "module" => "Elixir.MetadataLogger",
      "pid" => "#PID<" <> _ = inspect(self()),
      "gl" => "#PID<" <> _ = inspect(self()),
      "ancestors" => ["#PID<" <> _ = inspect(self())],
      "callers" => ["#PID<" <> _ = inspect(self())],
      "domain" => ["elixir"],
      "metadata" => %{"foo" => "bar", "list" => [1, 2, 3]},
      "timestamp" => @ts_iso8601,
      "level" => "info",
      "message" => "hi"
    }

    got =
      parse_formatted(:info, "hi", @ts_tuple,
        module: MetadataLogger,
        function: "hello/1",
        file: "/my/file.ex",
        line: 11,
        pid: self(),
        gl: self(),
        foo: :bar,
        list: [1, 2, 3],
        domain: [:elixir],
        ancestors: [self()],
        callers: [self()]
      )

    assert expected == got
  end

  test "does not include nil value for known metadata" do
    expected = %{
      "metadata" => %{"nil_val" => nil, "empty_map" => %{}, "empty_list" => []},
      "timestamp" => @ts_iso8601,
      "level" => "info",
      "message" => "hi"
    }

    got = parse_formatted(:info, "hi", @ts_tuple, nil_val: nil, empty_map: %{}, empty_list: [])

    assert expected == got
  end

  test "handles supported types in metadata" do
    expected = %{
      "metadata" => %{
        "atom" => "foo",
        "float" => 0.12345678,
        "int" => 1,
        "list" => [1, "1"],
        "map" => %{"a" => 1, "b" => "2"},
        "string" => "1"
      },
      "timestamp" => @ts_iso8601,
      "level" => "info",
      "message" => "hi"
    }

    got =
      parse_formatted(:info, "hi", @ts_tuple,
        atom: :foo,
        float: 0.12345678,
        int: 1,
        list: [1, "1"],
        map: %{"a" => 1, :b => "2"},
        string: "1"
      )

    assert expected == got
  end

  test "handle unsupported types in metadata" do
    cases = [
      bitstring: <<1::3>>,
      fn: fn -> nil end,
      pid: self(),
      struct: %URI{},
      tuple: {}
    ]

    Enum.each(cases, fn {_key, val} ->
      output = parse_formatted(:info, "hi", @ts_tuple, val: val)
      expected_metadata = inspect(val: val)
      expected_ts = inspect(@ts_tuple)

      assert %{
               "level" => "error",
               "message" => "could not format json message",
               "logger_data" => %{
                 "level" => ":info",
                 "message" => "\"hi\"",
                 "ts" => ^expected_ts,
                 "exception" => exception,
                 "metadata" => ^expected_metadata
               }
             } = output

      assert String.starts_with?(exception, "%Protocol.UndefinedError{")
    end)
  end

  test "handles erlang error logger metadata" do
    expected = %{
      "crash_reason" => "{:foo, []}",
      "initial_call" => "{:hello, :world, 1}",
      "registered_name" => ":me",
      "metadata" => %{},
      "timestamp" => @ts_iso8601,
      "level" => "info",
      "message" => "hi"
    }

    got =
      parse_formatted(:info, "hi", @ts_tuple,
        crash_reason: {:foo, []},
        initial_call: {:hello, :world, 1},
        report_cb: & &1,
        mfa: {Foo, :bar, 1},
        registered_name: :me,
        error_logger: %{
          report_cb: & &1,
          tag: :info_report,
          type: :std_info
        }
      )

    assert expected == got
  end

  test "Always appends a newline" do
    assert String.ends_with?(formatted(:info, "hi", @ts_tuple, %{}), "\n")
    assert String.ends_with?(formatted(:info, "hi", @ts_tuple, val: %URI{}), "\n")
  end

  defp formatted(level, message, ts, metadata) do
    output_iodata = MetadataLogger.format(level, message, ts, metadata)

    IO.iodata_to_binary(output_iodata)
  end

  defp parse_formatted(level, message, ts, metadata) do
    output = formatted(level, message, ts, metadata)

    Jason.decode!(output)
  end
end


test/metadata_logger_json_formatter_test.exs
defmodule MetadataLoggerJsonFormatterTest do
  use ExUnit.Case
  doctest MetadataLoggerJsonFormatter

  require Logger

  import ExUnit.CaptureLog

  setup do
    on_exit(fn ->
      :ok =
        Logger.configure_backend(
          :console,
          format: nil,
          device: :user,
          level: nil,
          metadata: :all,
          colors: [enabled: false]
        )
    end)

    Logger.configure_backend(:console,
      format: {MetadataLoggerJsonFormatter, :format},
      colors: [enabled: false],
      truncate: :infinity,
      metadata: :all
    )

    :ok
  end

  test "log in json" do
    assert %{
             "module" => "Elixir.MetadataLoggerJsonFormatterTest",
             "pid" => _,
             "metadata" => %{"foo" => "bar", "list" => [1, 2, 3]},
             "timestamp" => _,
             "level" => "info",
             "message" => "hi"
           } = parsed_log(:info, "hi", foo: :bar, list: [1, 2, 3])
  end

  test "log metadata in current process" do
    try do
      Logger.metadata(foo: :bar)
      assert %{"metadata" => %{"foo" => "bar"}} = parsed_log(:info, "hi", [])
    after
      Logger.metadata(foo: nil)
    end
  end

  test "handles missing Logger metadata from Logger.bare_log/3" do
    keys_got =
      capture_log(fn -> Logger.bare_log(:info, "hello", hello: :world) end)
      |> Jason.decode!()
      |> Map.keys()
      |> MapSet.new()

    known_keys =
      [
        :app,
        :module,
        :function,
        :file,
        :line
        # :pid,
      ]
      |> Enum.map(&to_string/1)
      |> MapSet.new()

    assert MapSet.new() == MapSet.intersection(keys_got, known_keys)
  end

  test "handles supported types in metadata" do
    cases = [
      atom: [:foo, "foo"],
      float: [0.12345678, 0.12345678],
      int: [1, 1],
      list: [[1, "1"], [1, "1"]],
      map: [%{"a" => 1, :b => "2"}, %{"a" => 1, "b" => "2"}],
      string: ["1", "1"]
    ]

    Enum.each(cases, fn {key, [val, expected]} ->
      assert {^key, %{"metadata" => %{"val" => ^expected}}} =
               {key, parsed_log(:info, "hello", val: val)}
    end)
  end

  test "handle unsupported types in metadata" do
    cases = [
      bitstring: <<1::3>>,
      fn: fn -> nil end,
      pid: self(),
      struct: %URI{},
      tuple: {}
    ]

    Enum.each(cases, fn {key, val} ->
      output = capture_log(fn -> Logger.log(:info, "hello", val: val) end)

      assert String.starts_with?(output, "could not format: %Protocol.UndefinedError"),
             "should not handle #{inspect(val)} (#{key}); #{output}"
    end)
  end

  defp parsed_log(level, message, metadata) do
    captured = capture_log(fn -> Logger.log(level, message, metadata) end)
    Jason.decode!(captured)
  end
end


mix.exs
defmodule MetadataLogger.MixProject do
  use Mix.Project

  @version "0.3.2"

  def project do
    [
      app: :metadata_logger,
      version: @version,
      elixir: "~> 1.6",
      start_permanent: Mix.env() == :prod,
      deps: deps(),

      # hex
      description: "Logging with metadata",
      package: package(),

      # ex_doc
      docs: docs()
    ]
  end

  def application do
    [
      extra_applications: []
    ]
  end

  defp deps do
    [
      {:jason, "~> 1.0"},
      {:credo, "~> 1.1.0", only: [:dev, :test], runtime: false},
      {:ex_doc, "~> 0.19", only: :dev, runtime: false}
    ]
  end

  defp package do
    [
      licenses: ["MIT"],
      links: %{
        "GitHub" => "https://github.com/elixir-metadata-logger/metadata_logger",
        "Changelog" =>
          "https://github.com/elixir-metadata-logger/metadata_logger/blob/main/CHANGELOG.md"
      },
      maintainers: ["Chulki Lee"]
    ]
  end

  defp docs do
    [
      name: "MetadataLogger",
      source_ref: "v#{@version}",
      canonical: "https://hexdocs.pm/metadata_logger",
      source_url: "https://github.com/elixir-metadata-logger/metadata_logger"
    ]
  end
end


config/config.exs
# This file is responsible for configuring your application
# and its dependencies with the aid of the Mix.Config module.
use Mix.Config

# This configuration is loaded before any dependency and is restricted
# to this project. If another project depends on this project, this
# file won't be loaded nor affect the parent project. For this reason,
# if you want to provide default values for your application for
# 3rd-party users, it should be done in your "mix.exs" file.

# You can configure your application as:
#
#     config :metadata_logger_json_formatter, key: :value
#
# and access this configuration in your application as:
#
#     Application.get_env(:metadata_logger_json_formatter, :key)
#
# You can also configure a 3rd-party app:
#
#     config :logger, level: :info
#

# It is also possible to import configuration files, relative to this
# directory. For example, you can emulate configuration per environment
# by uncommenting the line below and defining dev.exs, test.exs and such.
# Configuration from the imported file will override the ones defined
# here (which is why it is important to import them last).
#
#     import_config "#{Mix.env}.exs"


lib/metadata_logger_json_formatter.ex
defmodule MetadataLoggerJsonFormatter do
  @moduledoc """
  Logger formatter to print message and metadata in a single-line json.

  ```json
  config :logger, :console,
    level: :debug,
    format: {MetadataLoggerJsonFormatter, :format},
    colors: [enabled: false],
    metadata: :all,
    truncate: :infinity,
    utc_log: true
  ```
  """
  def format(level, message, ts, metadata) do
    line =
      metadata
      |> build_line()
      |> scrub(level)
      |> Map.put(:timestamp, format_timestamp(ts))
      |> Map.put(:level, level)
      |> Map.put(:message, to_string(message))
      |> Jason.encode_to_iodata!()

    [line, "\n"]
  rescue
    e -> "could not format: #{inspect(e)} - #{inspect({level, ts, message, metadata})}"
  end

  defp format_timestamp({{y, month, d}, {h, minutes, s, mil}}) do
    {:ok, dt} = NaiveDateTime.new(y, month, d, h, minutes, s, mil)
    NaiveDateTime.to_iso8601(dt)
  end

  defp build_line(metadata) do
    with m <- Enum.into(metadata, %{}),
         {app, m} <- Map.pop(m, :application),
         {module, m} <- Map.pop(m, :module),
         {function, m} <- Map.pop(m, :function),
         {file, m} <- Map.pop(m, :file),
         {line, m} <- Map.pop(m, :line),
         {pid, m} <- Map.pop(m, :pid) do
      %{metadata: m}
      |> put_val(:app, app)
      |> put_val(:module, module)
      |> put_val(:function, function)
      |> put_val(:file, file)
      |> put_val(:line, line)
      |> put_val(:pid, nil_or_inspect(pid))
    end
  end

  defp nil_or_inspect(nil), do: nil
  defp nil_or_inspect(val), do: inspect(val)

  def put_val(map, _key, nil), do: map
  def put_val(map, key, val), do: Map.put(map, key, val)

  defp scrub(map, _level) do
    map
    |> Map.delete(:func)
    |> Map.delete(:file)
    |> Map.delete(:line)
  end
end


lib/metadata_logger.ex
defmodule MetadataLogger do
  @moduledoc """
  Logging with metadata.

  ## Configuration

  ```elixir
  config :logger, :console,
    level: :debug,
    format: {MetadataLogger, :format},
    colors: [enabled: false],
    metadata: :all,
    truncate: :infinity,
    utc_log: true
  ```
  """

  @doc """
  Formatter function to print message and metadata in a single-line json.

  Make sure all metadata except known metadata encodable by `Jason.encode_to_iodata!/2`.
  See `log_to_map/4` for transformation on known metadata.

  It removes `:function`, `:file`, and `:line` from the log.

  ## Examples

      iex> MetadataLogger.format(
      ...>   :info,
      ...>   ["hello", " ", "world"],
      ...>   {{2019, 11, 22}, {12, 23, 45, 678}},
      ...>   function: "hello/1",
      ...>   file: "/my/file.ex",
      ...>   line: 11,
      ...>   foo: :bar
      ...> ) |> Jason.decode!()
      %{
        "level" => "info",
        "message" => "hello world",
        "metadata" => %{"foo" => "bar"},
        "timestamp" => "2019-11-22T12:23:45.000678"
      }

      iex> MetadataLogger.format(
      ...>   :info,
      ...>   %{hello: :world},
      ...>   {{2019, 11, 22}, {12, 23, 45, 678}},
      ...>   function: "hello/1",
      ...>   file: "/my/file.ex",
      ...>   line: 11,
      ...>   foo: :bar
      ...> ) |> Jason.decode!()
      %{
        "level" => "info",
        "message" => %{"hello" => "world"},
        "metadata" => %{"foo" => "bar"},
        "timestamp" => "2019-11-22T12:23:45.000678"
      }

  """
  @spec format(Logger.level(), Logger.message(), Logger.Formatter.time(), keyword) ::
          IO.chardata()
  def format(level, message, ts, metadata) do
    map = log_to_map(level, message, ts, metadata)

    [
      map
      |> scrub()
      |> Jason.encode_to_iodata!(map),
      "\n"
    ]
  rescue
    e ->
      [
        %{
          level: "error",
          message: "could not format json message",
          logger_data: %{
            exception: inspect(e),
            level: inspect(level),
            ts: inspect(ts),
            message: inspect(message),
            metadata: inspect(metadata)
          }
        }
        |> Jason.encode!(),
        "\n"
      ]
  end

  @doc """
  Get a map from log formatter arguments.

  It converts `t:Logger.Formatter.time/0` into `NaiveDateTime` so it is recommended to configure logger to use UTC by setting `:utc_log` to `true`.

  It converts metadata keyword into map using `Enum.into/2` therefore duplicated keys will be removed.

  It moves following known metadata to the top level. See [Logger Metadata](https://hexdocs.pm/logger/Logger.html#module-metadata) for details.

  - `:application`
  - `:module`
  - `:function`
  - `:file`
  - `:line`
  - `:pid`
  - `:gl`
  - `:crash_reason`
  - `:initial_call`
  - `:registered_name`
  - `:domain`
  - `:ancestors`
  - `:callers`

  Followings metadata will be removed:

  - `:mfa`: see `:module` and `:funtion`
  - `:report_cb`

  ## Examples

      iex> MetadataLogger.log_to_map(
      ...>   :info,
      ...>   ["hello", " ", "world"],
      ...>   {{2019, 11, 22}, {12, 23, 45, 678}},
      ...>   foo: :bar
      ...> )
      %{
        level: :info,
        message: "hello world",
        metadata: %{foo: :bar},
        timestamp: ~N[2019-11-22 12:23:45.000678]
      }

      iex> MetadataLogger.log_to_map(
      ...>   :info,
      ...>   %{hello: :world},
      ...>   {{2019, 11, 22}, {12, 23, 45, 678}},
      ...>   []
      ...> )
      %{
        level: :info,
        message: %{hello: :world},
        metadata: %{},
        timestamp: ~N[2019-11-22 12:23:45.000678]
      }

      iex> MetadataLogger.log_to_map(
      ...>   :info,
      ...>   [foo: 1, foo: 2],
      ...>   {{2019, 11, 22}, {12, 23, 45, 678}},
      ...>   []
      ...> )
      %{
        level: :info,
        message: %{foo: 2},
        metadata: %{},
        timestamp: ~N[2019-11-22 12:23:45.000678]
      }

  """
  @spec log_to_map(
          Logger.level(),
          Logger.message() | Logger.report(),
          Logger.Formatter.time(),
          keyword
        ) ::
          map()
  def log_to_map(level, message, ts, metadata) do
    with m <- Enum.into(metadata, %{}),
         m <- Map.drop(m, [:error_logger, :mfa, :report_cb]),
         {app, m} <- Map.pop(m, :application),
         {module, m} <- Map.pop(m, :module),
         {function, m} <- Map.pop(m, :function),
         {file, m} <- Map.pop(m, :file),
         {line, m} <- Map.pop(m, :line),
         {pid, m} <- Map.pop(m, :pid),
         {gl, m} <- Map.pop(m, :gl),
         {ancestors, m} <- Map.pop(m, :ancestors),
         {callers, m} <- Map.pop(m, :callers),
         {crash_reason, m} <- Map.pop(m, :crash_reason),
         {initial_call, m} <- Map.pop(m, :initial_call),
         {domain, m} <- Map.pop(m, :domain),
         {registered_name, m} <- Map.pop(m, :registered_name) do
      %{metadata: m}
      |> put_val(:app, app)
      |> put_val(:module, module)
      |> put_val(:function, function)
      |> put_val(:file, file)
      |> put_val(:line, line)
      |> put_val(:pid, nil_or_inspect(pid))
      |> put_val(:gl, nil_or_inspect(gl))
      |> put_val(:crash_reason, nil_or_inspect(crash_reason))
      |> put_val(:initial_call, nil_or_inspect(initial_call))
      |> put_val(:registered_name, nil_or_inspect(registered_name))
      |> put_val(:domain, domain)
      |> put_val(:ancestors, nil_or_inspect_list(ancestors))
      |> put_val(:callers, nil_or_inspect_list(callers))
    end
    |> Map.put(:level, level)
    |> Map.put(:message, transform_message(message))
    |> Map.put(:timestamp, transform_timestamp(ts))
  end

  defp nil_or_inspect(nil), do: nil
  defp nil_or_inspect(val), do: inspect(val)

  defp nil_or_inspect_list(nil), do: nil
  defp nil_or_inspect_list(val), do: Enum.map(val, &inspect/1)

  defp put_val(map, _key, nil), do: map
  defp put_val(map, key, val), do: Map.put(map, key, val)

  defp transform_message(%_{} = m), do: to_string(m)
  defp transform_message(m) when is_map(m), do: m
  defp transform_message([{_k, _v} | _t] = m), do: Enum.into(m, %{})
  defp transform_message(m) when is_list(m), do: IO.iodata_to_binary(m)
  defp transform_message(m), do: to_string(m)

  defp transform_timestamp({{y, month, d}, {h, minutes, s, mil}}) do
    {:ok, dt} = NaiveDateTime.new(y, month, d, h, minutes, s, mil)
    dt
  end

  defp scrub(map) do
    map
    |> Map.delete(:function)
    |> Map.delete(:file)
    |> Map.delete(:line)
  end
end
